cmake/onnxruntime_training.cmake

```diff
@@ -230,4 +231,47 @@ if (onnxruntime_BUILD_UNIT_TESTS)
   target_link_libraries(onnxruntime_training_gpt2 PRIVATE onnxruntime_training_runner onnxruntime_training ${ONNXRUNTIME_LIBS} ${onnxruntime_EXTERNAL_LIBRARIES})
   set_target_properties(onnxruntime_training_gpt2 PROPERTIES FOLDER "ONNXRuntimeTest")
 
+  # Training API Tests
+  # Currently disable it by default for internal development usage.
+  if (onnxruntime_ENABLE_TRAINING_ON_DEVICE)
+    # Only files in the trainer and common folder will be compiled into test trainer.
+    file(GLOB_RECURSE training_api_test_trainer_src
+      "${ORTTRAINING_SOURCE_DIR}/test/training_api/common/*.cc"
+      "${ORTTRAINING_SOURCE_DIR}/test/training_api/common/*.h"
+      "${ORTTRAINING_SOURCE_DIR}/test/training_api/trainer/*.cc"
+      "${ORTTRAINING_SOURCE_DIR}/test/training_api/trainer/*.h"
+    )
+    onnxruntime_add_executable(onnxruntime_test_trainer ${training_api_test_trainer_src})
+
+    if(UNIX AND NOT APPLE)
+      if (HAS_NO_MAYBE_UNINITIALIZED)
```

Comment:
ah, I missed that one in onnxruntime_training.cmake.

I think the way `HAS_NO_MAYBE_UNINITIALIZED` is set has an issue. for GCC, `check_cxx_compiler_flag(-Wno-<warning>)` may indicate that a warning is valid even though it's not, since GCC allows unrecognized warnings in the `-Wno-` form (more info [here](https://gcc.gnu.org/onlinedocs/gcc-12.1.0/gcc/Warning-Options.html)). this is not the case for the `-W<warning>` form. so using `HAS_MAYBE_UNINITIALIZED` would be correct.

as a follow up question, why do we need to disable this warning? might be good have a comment with the reason.

Reply:
> as a follow up question, why do we need to disable this warning? might be good have a comment with the reason.

I actually don't know the answer to this question. @pengwa do you know why we need to disable this waning? My guess is that this was done historically for some reason, and it was perpetuated as more targets were added.

In this PR however, I don't think we need to check for this. I think it is safe for me to remove for this PR.